Use spalloc to get a SpiNNaker machine

mundya opened the pull request:
- Use [spalloc](https://github.com/project-rig/spalloc) to get automatically allocated SpiNNaker boards
- Allow the Simulator be called with the hostname of the SpiNNaker board to use
- Use the OS to get a port for incoming packets from SpiNNaker

@pabogdan - I believe that the latter two points are things you also did at some point?

mundya committed:
Specify SpiNNaker machine in simulator init

Allow specification of the SpiNNaker machine when initialising a
Simulator object. A SpiNNaker machine allocated by `spalloc` can be used
with:

```python
from spalloc import Job

with Job(3) as j:
    sim = Simulator(model, hostname=j.hostname)
    ...
```

mundya committed:
Automatically allocate port for Node communication

Allow the operating system to select a free port for communication with
a SpiNNaker machine. This may require some changes to firewall settings.

mundya committed:
Defer booting the SpiNNaker machine

Don't construct a machine controller or boot the SpiNNaker machine until
the netlist has been constructed; this will make automatic provision of
SpiNNaker machines with `spalloc` simpler.

mundya committed:
Allow automatic allocation of a SpiNNaker machine

Add a `use_spalloc` argument to the Simulator. If `use_spalloc` is True
then `spalloc` is used to allocate a SpiNNaker machine to perform the
simulation.

mossblaser commented on nengo_spinnaker/simulator.py:
```diff
@@ -126,6 +126,36 @@ def __init__(self, network, dt=0.001, period=10.0, timescale=1.0):
+            # per chip and 48 chips per board).
+            n_cores = sum(v.resources.get(Cores, 0) for v in
+                          self.netlist.vertices)
+            n_boards = int(np.ceil((n_cores / 16.) / 48.))
```
I'd make a fudge factor a parameter to this with a generous default. You'll want to over-allocate on boards due to boards not _always_ having 48 chips and packing resulting in non-optimal placement.

mundya replied:
I wondered about that, any factor suggestions? 1.5? 2?

mossblaser replied:
Not got much intuition here... Packing problems will be your main enemy... Seems like that part of this fudge factor could be estimated from the P&R inputs somehow... A new rig feature?

For the dead chips thing, chips are generally missing at a rate of just above 1 in 100 so 1.1 would be a good safe fudge factor for that.

mossblaser commented on nengo_spinnaker/simulator.py:
```diff
@@ -126,6 +126,36 @@ def __init__(self, network, dt=0.001, period=10.0, timescale=1.0):
+            n_boards = int(np.ceil((n_cores / 16.) / 48.))
+
+            # Request the job
+            self.job = Job(n_boards)
```
I'd print the Job ID in the logging data at this point.

mossblaser commented on nengo_spinnaker/simulator.py:
```diff
@@ -126,6 +126,36 @@ def __init__(self, network, dt=0.001, period=10.0, timescale=1.0):
+            self.job.wait_until_ready()
+
+            # Store the hostname
+            hostname = self.job.hostname
```
I'd print this hostname (and probably also the machine-name, and number of boards) at this point.

mossblaser commented on nengo_spinnaker/simulator.py:
```diff
@@ -126,6 +126,36 @@ def __init__(self, network, dt=0.001, period=10.0, timescale=1.0):
+            hostname = self.job.hostname
+
+        self.controller = MachineController(hostname)
+        self.controller.boot()
```
Frustratingly, for whatever reason, in most Spalloc'd systems you need to wait a bit before actually attempting to boot an allocated machine. This appears to be an artefact of having more layers of network between you and the machine than usual causing ARP to fail or something... Not really got a good solution to this... In future Spalloc server versions a longer pause before reporting a board as up may be the solution...

mundya replied:
Any idea what kind of delay to insert before calling boot? Should a second do it?

mossblaser replied:
A second often is sufficient but 5 always seems to be... I'd err on the side of having too many...

mossblaser commented on nengo_spinnaker/simulator.py:
```diff
@@ -341,6 +371,10 @@ def close(self):
 
+            # Destroy the job if we allocated one
+            if self.job is not None:
+                self.job.destroy()
```
How reliably will the surrounding 'close' function be called out of curiosity?

mundya replied:
It will work for all cases that [`atexit`](https://docs.python.org/3.5/library/atexit.html) does; in my experience it is _always_ called.

mossblaser replied:
Superb!

mundya committed:
Improvements to spalloc usage

Thanks @mossblaser

mundya committed:
Increase pre-boot delay for spalloc

mundya commented on nengo_spinnaker/simulator.py:
```diff
@@ -126,6 +137,45 @@ def __init__(self, network, dt=0.001, period=10.0, timescale=1.0):
+            # spalloc recommends a slight delay before attempting to boot the
+            # machine, later versions of spalloc server may relax this
+            # requirement.
+            time.sleep(1)
```
1s is too short

mundya commented on nengo_spinnaker/simulator.py:
```diff
@@ -61,16 +63,25 @@ def __init__(self, network, dt=0.001, period=10.0, timescale=1.0):
+            Hostname of the SpiNNaker machine to use; if None then the machine
+            specified in the config file will be used.
+        use_spalloc : bool
+            Allocate a SpiNNaker machine for the simulator using ``spalloc``.
```
Should be `None` by default with the final value selected from the config file.

mundya committed:
Allow setting of use_spalloc in config file

Adding `use_spalloc: True` or similar to the nengo_spinnaker
configuration file will default nengo_spinnaker to using a
spalloc-allocated machine.

Some slight changes made to logging messages.

mundya committed:
Report number of boards used, not number requested

Update the logging information to report the number of boards allocated
by spalloc rather than the number that were requested.

mundya committed:
Changes for larger machines.

Thanks @neworderofjamie and @mossblaser.

mossblaser commented on nengo_spinnaker/simulator.py:
```diff
@@ -66,8 +66,9 @@ def __init__(self, network, dt=0.001, period=10.0, timescale=1.0,
-        use_spalloc : bool
+        use_spalloc : bool or None
             Allocate a SpiNNaker machine for the simulator using ``spalloc``.
+            If None then the setting specified in the config file will be used.
```
Worth reiterating what the default when not specified in the config is?

mundya replied:
Good point, thanks

mossblaser commented:
These changes LGTM!
